[SPARK-21503][UI]: Spark UI shows incorrect task status for a killed Executor Process

pgandhi999 opened the pull request:
The executor tab on Spark UI page shows task as completed when an executor process that is running that task is killed using the kill command.
Added the case ExecutorLostFailure which was previously not there, thus, the default case would be executed in which case, task would be marked as completed. This case will consider all those cases where executor connection to Spark Driver was lost due to killing the executor process, network connection etc.

## How was this patch tested?
Manually Tested the fix by observing the UI change before and after.
Before:
<img width="1398" alt="screen shot-before" src="https://user-images.githubusercontent.com/22228190/28482929-571c9cea-6e30-11e7-93dd-728de5cdea95.png">
After:
<img width="1385" alt="screen shot-after" src="https://user-images.githubusercontent.com/22228190/28482964-8649f5ee-6e30-11e7-91bd-2eb2089c61cc.png">


Please review http://spark.apache.org/contributing.html before opening a pull request.

Commit:
[SPARK-21503]: Fixed the issue

Added the case ExecutorLostFailure which was previously not there, thus, the default case would be executed in which case, task would be marked as completed.

AmplabJenkins commented:
Can one of the admins verify this patch?

tgravescs commented:
ok to test

SparkQA commented:
**[Test build #79850 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79850/testReport)** for PR 18707 at commit [`172fc20`](https://github.com/apache/spark/commit/172fc20898896058b7288360eb5292ed9df9d79c).

zsxwing commented on core/src/main/scala/org/apache/spark/ui/exec/ExecutorsTab.scala:
```diff
@@ -140,6 +140,8 @@ class ExecutorsListener(storageStatusListener: StorageStatusListener, conf: Spar
           return
         case _: ExceptionFailure =>
           taskSummary.tasksFailed += 1
+        case _: ExecutorLostFailure =>
```
Looks like we can use `info.successful`?

jiangxb1987 replied:
+1

pgandhi999 replied:
Hi,
   I have replaced most of the task failed cases with info.successful. Tested it as well. Have also written unit test to simulate my issue. Thank you.

SparkQA commented:
**[Test build #79850 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79850/testReport)** for PR 18707 at commit [`172fc20`](https://github.com/apache/spark/commit/172fc20898896058b7288360eb5292ed9df9d79c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79850/
Test PASSed.

Commit:
[SPARK-21503][UI]: Adding changes as per comments

SparkQA commented:
**[Test build #79937 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79937/testReport)** for PR 18707 at commit [`81422e0`](https://github.com/apache/spark/commit/81422e0f634c0f06eb2ea29fba4281176a1ab528).

SparkQA commented:
**[Test build #79937 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79937/testReport)** for PR 18707 at commit [`81422e0`](https://github.com/apache/spark/commit/81422e0f634c0f06eb2ea29fba4281176a1ab528).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79937/
Test PASSed.

pgandhi999 committed:
Merge pull request #1 from apache/master

Apache Spark Pull Request - July 26, 2017

SparkQA commented:
**[Test build #79976 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79976/testReport)** for PR 18707 at commit [`55c6c37`](https://github.com/apache/spark/commit/55c6c37d09b41ae6914edb5d067e7f2c252ac92a).

Commit:
[SPARK-21541]: Spark Logs show incorrect job status for a job that does not create SparkContext

Added a flag to check whether user has initialized Spark Context. If it is true, then we let Application Master unregister with Resource Manager else we do not.

SparkQA commented:
**[Test build #79977 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79977/testReport)** for PR 18707 at commit [`f454c89`](https://github.com/apache/spark/commit/f454c8933e07967548095e068063bd313ae4845c).

SparkQA commented:
**[Test build #79977 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79977/testReport)** for PR 18707 at commit [`f454c89`](https://github.com/apache/spark/commit/f454c8933e07967548095e068063bd313ae4845c).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79977/
Test FAILed.

Commit:
Revert "[SPARK-21541]: Spark Logs show incorrect job status for a job that does not create SparkContext"

This reverts commit f454c8933e07967548095e068063bd313ae4845c.

"Merged another issue to this one by mistake"

SparkQA commented:
**[Test build #79978 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79978/testReport)** for PR 18707 at commit [`6b7d5c6`](https://github.com/apache/spark/commit/6b7d5c6e2565c7c4dd97f31fe404c59e73c7474c).

SparkQA commented:
**[Test build #79976 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79976/testReport)** for PR 18707 at commit [`55c6c37`](https://github.com/apache/spark/commit/55c6c37d09b41ae6914edb5d067e7f2c252ac92a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `        throw new AnalysisException(s\"UDF class $className doesn't implement any UDF interface\")`
  * `        throw new AnalysisException(s\"It is invalid to implement multiple UDF interfaces, UDF class $className\")`
  * `              throw new AnalysisException(s\"UDF class with $n type arguments is not supported.\")`
  * `            throw new AnalysisException(s\"Can not instantiate class $className, please make sure it has public non argument constructor\")`
  * `      case e: ClassNotFoundException => throw new AnalysisException(s\"Can not load class $className, please make sure it is on the classpath\")`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79976/
Test PASSed.

SparkQA commented:
**[Test build #79978 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79978/testReport)** for PR 18707 at commit [`6b7d5c6`](https://github.com/apache/spark/commit/6b7d5c6e2565c7c4dd97f31fe404c59e73c7474c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79978/
Test PASSed.

jiangxb1987 commented on core/src/main/scala/org/apache/spark/ui/exec/ExecutorsTab.scala:
```diff
@@ -138,10 +138,12 @@ class ExecutorsListener(storageStatusListener: StorageStatusListener, conf: Spar
           // could have failed half-way through. The correct fix would be to keep track of the
           // metrics added by each attempt, but this is much more complicated.
           return
```
nit:
```
if (taskEnd.reason == Resubmitted) {
  return ;
}
```

pgandhi999 replied:
I have made the required changes and verified them. Thank you.

jiangxb1987 commented on core/src/main/scala/org/apache/spark/ui/exec/ExecutorsTab.scala:
```diff
@@ -138,10 +138,12 @@ class ExecutorsListener(storageStatusListener: StorageStatusListener, conf: Spar
+      if (info.successful) {
+        taskSummary.tasksComplete += 1
+      } else {
+        taskSummary.tasksFailed += 1
```
I'm not sure this keeps the previous behavior, cc @zsxwing

pgandhi999 replied:
I went through the list of possible task failure reasons, and I found info.successful to be updated for all except Resubmitted case. If you think there is a possibility of any other reason that has been overlooked by me, do let me know.

pgandhi999 committed:
Merge pull request #2 from apache/master

Spark - July 28, 2017

SparkQA commented:
**[Test build #80026 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80026/testReport)** for PR 18707 at commit [`bc41664`](https://github.com/apache/spark/commit/bc4166490d2ff68898c00fae4c1ca1b8abe1e795).

Commit:
[SPARK-21503]- Making Changes as per comments

[SPARK-21503]- Making Changes as per comments: Removed match case statement and replaced it with an if clause.

SparkQA commented:
**[Test build #80027 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80027/testReport)** for PR 18707 at commit [`e46126f`](https://github.com/apache/spark/commit/e46126fe0f3d8d6f92f7f51c30d8c2154bddc126).

SparkQA commented:
**[Test build #80026 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80026/testReport)** for PR 18707 at commit [`bc41664`](https://github.com/apache/spark/commit/bc4166490d2ff68898c00fae4c1ca1b8abe1e795).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class OneVsRestParams(HasFeaturesCol, HasLabelCol, HasWeightCol, HasPredictionCol):`
  * `class ArrowWriter(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80026/
Test PASSed.

SparkQA commented:
**[Test build #80027 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80027/testReport)** for PR 18707 at commit [`e46126f`](https://github.com/apache/spark/commit/e46126fe0f3d8d6f92f7f51c30d8c2154bddc126).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80027/
Test PASSed.

umehrot2 commented on core/src/test/scala/org/apache/spark/ui/StagePageSuite.scala:
```diff
@@ -84,6 +84,19 @@ class StagePageSuite extends SparkFunSuite with LocalSparkContext {
         jobListener.onTaskEnd(
           SparkListenerTaskEnd(0, 0, "result", Success, taskInfo, taskMetrics))
     }
+    (3 to 4).foreach {
```
I don't see a test case which uses this method to test that the task is correctly displayed as failed. In that case, is this test suite modification needed ?

pgandhi999 replied:
You are right. I did try to simulate a similar unit test for Executors Tab UI, but it needs creating a SparkUI object which cannot be accessed from StagePageSuite. Hence, I am reverting my unit test for the time being.

pgandhi999 committed:
Merge pull request #4 from apache/master

Spark - August 1, 2017

Commit:
[SPARK-21503]: Reverting Unit Test Code

[SPARK-21503]: Reverting Unit Test Code - Not needed.

SparkQA commented:
**[Test build #80122 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80122/testReport)** for PR 18707 at commit [`7f03341`](https://github.com/apache/spark/commit/7f03341093c843086920e8218463b5d2ba6e37d2).

SparkQA commented:
**[Test build #80119 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80119/testReport)** for PR 18707 at commit [`9b3cebc`](https://github.com/apache/spark/commit/9b3cebc6b65d2da835f02efaa27015cfd1b0ccae).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnresolvedSubqueryColumnAliases(`
  * `sealed trait FrameType `
  * `sealed trait SpecialFrameBoundary extends Expression with Unevaluable `
  * `sealed trait WindowFrame extends Expression with Unevaluable `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80119/
Test PASSed.

SparkQA commented:
**[Test build #80122 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80122/testReport)** for PR 18707 at commit [`7f03341`](https://github.com/apache/spark/commit/7f03341093c843086920e8218463b5d2ba6e37d2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80122/
Test PASSed.

jiangxb1987 commented:
LGTM, cc @zsxwing @cloud-fan

cloud-fan commented on core/src/main/scala/org/apache/spark/ui/exec/ExecutorsTab.scala:
```diff
@@ -131,17 +131,17 @@ class ExecutorsListener(storageStatusListener: StorageStatusListener, conf: Spar
+      if (taskEnd.reason == Resubmitted) {
+        return
+      }
+      if (info.successful) {
```
what's the relationship between `info.successful` and `taskEnd.reason`?

pgandhi999 replied:
def successful: Boolean = finished && !failed && !killed

So if  the task state shows finished and it has neither failed nor been killed, then info.successful is true. Now, there could be multiple reasons for a failed or killed task, and taskEnd.reason lists those. However, if task state is SUCCESS, so is the reason and thus, info.successful will be true in case of SUCCESS and false for all the other cases.

cloud-fan commented on core/src/main/scala/org/apache/spark/ui/exec/ExecutorsTab.scala:
```diff
@@ -131,17 +131,17 @@ class ExecutorsListener(storageStatusListener: StorageStatusListener, conf: Spar
-          // could have failed half-way through. The correct fix would be to keep track of the
-          // metrics added by each attempt, but this is much more complicated.
-          return
-        case _: ExceptionFailure =>
```
can we fix the problem by changing this line to `case _: TaskFailedReason`?

pgandhi999 replied:
So, I added info.successful check at the suggestion of @zsxwing and @jiangxb1987 . I feel this covers all the failed and killed cases which TaskFailedReason might have overlooked. However, if you feel otherwise, please let me know.

tgravescs commented:
change LGTM

@cloud-fan  I don't see any issues with the current approach, it was suggested earlier, but I think your suggestion would work also.  Functionally I'm not seeing any difference.  Did I miss something or are you ok with this way?

cloud-fan commented:
I was just asking about different fixes, I'm ok with the current fix.

thanks, merging to master/2.2!
